Added shell_sort in sort.py and remove errors in sort_tests.py

Aryamanz29 opened the pull request:
- Added shell sort in sort.py
-Added its test cases 
-Correct errors in sort_tests.py
closes #22

codesankalp commented on dsalgo/sort.py:
```diff
@@ -200,4 +202,27 @@ def partition(array, start, end):
+            y_index = y_index + 1
+        gap = gap//2
+    if reverse:
+        return arr[::-1]
```
styling errors fix that

Aryamanz29 replied:
Done!😊

codesankalp commented on dsalgo/sort.py:
```diff
@@ -200,4 +202,27 @@ def partition(array, start, end):
     return high
+def shell_sort(arr,reverse=False):
+    ''' Shell Sort
+        Complexity: O(n^2)
```
follow docstring method as mentioned.

Aryamanz29 replied:
Please review my changes.😊

devkapilbansal commented on dsalgo/sort.py:
```diff
@@ -1,9 +1,9 @@
 
+class Sort:
     def __new__(self, array, algo, reverse=False):
-        '''
```
Don't use double quotes here

Aryamanz29 replied:
> Don't use double quotes here

Right

devkapilbansal replied:
Don't use double quotes anywhere

devkapilbansal commented on dsalgo/sort.py:
```diff
@@ -18,169 +18,205 @@ def __new__(self, array, algo, reverse=False):
-            return selection(self.array,self.reverse)
-        if self.algo =='quick':
-            return quick(self.array,self.reverse)
+        if self.algo == "bubble_recursion":
+            return bubble_recursion(self.array, self.reverse)
+        if self.algo == "selection":
+            return selection(self.array, self.reverse)
+        if self.algo == "quick":
+            return quick(self.array, self.reverse)
+        if self.algo == "shell":
+            return shell_sort(self.array, self.reverse)
```
Write them in lexicographic order

Aryamanz29 replied:
> Write them in lexicographic order

Sure I'll correct this one.

devkapilbansal commented on dsalgo/sort.py:
```diff
@@ -18,169 +18,205 @@ def __new__(self, array, algo, reverse=False):
-    if reverse==True :
-     return array[::-1]
+
+        # Copy data to left[] and right[]
```
what is left and right? Use self-explanatory names as much as possible

Aryamanz29 replied:
> what is left and right? Use self-explanatory names as much as possible

I thought this is relevant because we splitting the array into halves so operations can be understood easily also there are single-line comments for further explanation.

devkapilbansal replied:
Then use names like left_half and write_half

devkapilbansal commented on dsalgo/sort.py:
```diff
@@ -18,169 +18,205 @@ def __new__(self, array, algo, reverse=False):
+                array[i] = array[i + 1]
+                array[i + 1] = num
+                bubble_recursion(array)
+        except IndexError:
```
You need to cover this error in the test case too

Aryamanz29 replied:
> You need to cover this error in the test case too

If it's got index error the tests already show that one and print that error on the terminal while testing, So I think we no need to cover that separately.

codesankalp replied:
No, it prints nothing on the terminal.
You can skip this by removing the try & except and changing the loop to iterate over n-1 elements.
Also, try-except also works fine.
There is no problem with that by me.

Aryamanz29 replied:
Added details for exception

devkapilbansal commented on dsalgo/sort.py:
```diff
@@ -23,164 +23,200 @@ def __new__(self, array, algo, reverse=False):
+
+def bubble_recursion(array, reverse=False):
+    for i, num in enumerate(array):
+        try:
```
Do we need try/except here

Commit:
Added shell_sort in sort.py and remove errors in sort_tests.py #22

-Implement shell_sort in sort.py
-Added tests and remove errors in sort_tests.py

Fixes #22

Aryamanz29 commented:
@devkapilbansal  pls review this PR again

devkapilbansal commented:
@Aryamanz29 add one more test case, like test_invalid_type

devkapilbansal commented:
Pinging @17sushmita @ravi5175 for review
